command/results_xunit.go

```diff
@@ -130,35 +129,60 @@ func (c *xunitResults) parseAndUploadResults(ctx context.Context, conf *model.Ta
 			return errors.Wrap(err, "error closing xunit file")
 		}
 
-		// go through all the tests
-		for _, suite := range testSuites {
-			for _, tc := range suite.TestCases {
-				// logs are only created when a test case does not succeed
-				test, log := tc.toModelTestResultAndLog(conf.Task)
-				if log != nil {
-					logs = append(logs, log)
-					logIdxToTestIdx = append(logIdxToTestIdx, len(tests))
-				}
-				tests = append(tests, test)
+		// parse tests and logs for this file
+		tests, logs, logIdxToTestIdx := generateLogsForOneFile(testSuites, conf.Task)
+
+		// send the logs to the API server and get log IDs back
+		for i, log := range logs {
+			if ctx.Err() != nil {
+				return errors.New("operation canceled")
 			}
-		}
-	}
 
-	td := client.TaskData{ID: conf.Task.Id, Secret: conf.Task.Secret}
+			logId, err := sendJSONLogs(ctx, logger, comm, td, log)
+			if err != nil {
+				catcher.Add(errors.Wrapf(err, "problem uploading logs for %s", log.Name))
+				continue
+			}
 
-	for i, log := range logs {
-		if ctx.Err() != nil {
-			return errors.New("operation canceled")
+			//store log IDs in the tests
+			index := logIdxToTestIdx[i]
+			if index >= 0 && index < len(tests) {
+				tests[index].LogId = logId
+				tests[index].LineNum = 1
+			}
 		}
 
-		logId, err := sendJSONLogs(ctx, logger, comm, td, log)
-		if err != nil {
-			logger.Task().Warningf("problem uploading logs for %s", log.Name)
-			continue
+		// send tests to API server
+		if err = sendJSONResults(ctx, conf, logger, comm, &task.TestResults{tests}); err != nil {
+			catcher.Add(errors.Wrapf(err, "problem uploading test data from file %s", reportFileLoc))
+		}
+	}
+
+	return catcher.Resolve()
+}
+
+func generateLogsForOneFile(testSuites []testSuite, t *task.Task) ([]task.TestResult, []*model.TestLog, map[int]int) {
+	tests := []task.TestResult{}
+	logs := []*model.TestLog{}
+	logIdxToTestIdx := make(map[int]int)
+	// go through all the tests
+	for _, suite := range testSuites {
+		for _, tc := range suite.TestCases {
+			// logs are only created when a test case does not succeed
+			test, log := tc.toModelTestResultAndLog(t)
+			tests = append(tests, test)
+			if log != nil {
+				if suite.SysOut != "" {
+					log.Lines = append(log.Lines, fmt.Sprintf("system-out: %s", suite.SysOut))
+				}
+				if suite.SysErr != "" {
+					log.Lines = append(log.Lines, fmt.Sprintf("system-err: %s", suite.SysErr))
+				}
+				logs = append(logs, log)
+				logIdxToTestIdx[len(logs)-1] = len(tests) - 1
```

Comment:
don't we want the map in the other direction? multiple tests could have logs of the same length?

Reply:
With the data structure used here, there is a 1:1 mapping of test struct to log struct (see line 172). A log:test mapping is more convenient for how it's used here